firedrake/variational_solver.py

```diff
@@ -300,7 +300,7 @@ def solve(self, bounds=None):
         coefficients = utils.unique(chain.from_iterable(form.coefficients() for form in forms if form is not None))
         # Make sure the solution dm is visited last
         solution_dm = self.snes.getDM()
-        problem_dms = [V.dm for V in utils.unique(c.function_space() for c in coefficients) if V.dm != solution_dm]
+        problem_dms = [V.dm for V in utils.unique(chain.from_iterable(c.function_space() for c in coefficients)) if V.dm != solution_dm]
```

Comment:
So that I understand better: The reason we need `chain.from_iterable` is because `c.function_space()` can be an iterable when dealing with mixed function spaces?

Reply:
Correct, the transfer manager is attached to an appctx on the dm of each individual field.